File: frontend/services/reality.service.ts
import { api } from "@/lib/axios";
import { API_ENDPOINTS } from "@/constants/api";
import type { RealityCompareResult, RealitySummary } from "@/types";

interface Envelope<T> {
  success: boolean;
  data: T;
}

class RealityService {
  /**
   * Compare live telemetry against the Digital Twin's last predicted
   * snapshot for a node. `telemetry` should carry at minimum
   * { cpu, memory, temperature } — the only keys the backend reads.
   */
  async compare(
    nodeId: string,
    telemetry: { cpu: number; memory: number; temperature: number }
  ): Promise<RealityCompareResult> {
    const response = await api.post<Envelope<RealityCompareResult>>(
      API_ENDPOINTS.reality.compare(nodeId),
      { telemetry }
    );
    return response.data.data;
  }

  /**
   * The backend's own `/reality/summary` is a hardcoded stub
   * (always {0,0,0}) — callers should prefer aggregating real
   * `compare()` results across nodes instead of relying on this.
   */
  async rawSummary(): Promise<RealitySummary> {
    const response = await api.get<Envelope<RealitySummary>>(
      API_ENDPOINTS.reality.summary
    );
    return response.data.data;
  }
}

export const realityService = new RealityService();

File: backend/app/services/reality_service.py
"""
Reality Service

Digital Twin Reality Verification Service
"""

from sqlalchemy.ext.asyncio import AsyncSession

from app.digital_twin.reality_checker import reality_checker


class RealityService:
    """
    Reality Verification Service
    """

    async def compare(
        self,
        db: AsyncSession,
        node_id: str,
        telemetry: dict,
    ):
        """
        Compare real telemetry with
        Digital Twin snapshot.
        """

        return reality_checker.compare(
            node_id=node_id,
            telemetry=telemetry,
        )

    async def summary(
        self,
        db: AsyncSession,
    ):
        """
        Reality Summary
        """

        return {
            "healthy_nodes": 0,
            "warning_nodes": 0,
            "critical_nodes": 0,
        }


reality_service = RealityService()